docs/recipes/reducers/06-NormalizingStateShape.md

```diff
@@ -0,0 +1,208 @@
+
+
+## Organizing Normalized Data in State
+
+A typical application will likely have a mixture of relational data and non-relational data.  While there is no single rule for exactly how those different types of data should be organized, one common pattern is to put the relational "tables" under a common parent key, such as "entities".  A state structure using this approach might look like:
+
+```js
+{
+    simpleDomainData1: {....},
+    simpleDomainData2: {....}
+    entities : {
+        entityType1 : {....},
+        entityType2 : {....}
+    }
+    ui : {
+        uiSection1 : {....},
+        uiSection2 : {....}
+    }
+}
+```
+
+This could be expanded in a number of ways.  For example, an application that does a lot of editing of entities might want to keep two sets of "tables" in the state, one for the "current" item values and one for the "work-in-progress" item values.  When an item is edited, its values could be copied into the "work-in-progress" section, and any actions that update it would be applied to the "work-in-progress" copy, allowing the editing form to be controlled by that set of data while another part of the UI still refers to the original version.  "Resetting" the edit form would simply require removing the item from the "work-in-progress" section and re-copying the original data from "current" to "work-in-progress", while "applying" the edits would involve copying the values from the "work-in-progress" section to the "current" section.
+
+
+## Relationships and Tables
+
+Because we're treating a portion of our Redux store as a "database", many of the principles of database design also apply here as well.  For example, if we have a many-to-many relationship, we can model that using an intermediate table that stores the IDs of the corresponding items (often known as a "join table" or an "associative table").  For consistency, we would probably also want to use the same `byId` and `allIds` approach that we used for the actual item tables, like this:
+
+```js
+{
+    entities: {
+        authors : { byId : {}, allIds : [] },
+        books : { byId : {}, allIds : [] },
+        authorBook : {
+            byId : {
+                1 : {
+                    id : 1,
+                    authorId : 5,
+                    bookId : 22
+                },
+                2 : {
+                    id : 2,
+                    authorId : 5,
+                    bookId : 15,
+                }
+                3 : {
+                    id : 3,
+                    authorId : 42,
+                    bookId : 12
+                }
+            },
+            allIds : [1, 2, 3]
+        
+        }
+    }
+}
+```
+
+Behaviors like "Look up all books by this author" then become an O(n) operation, with a linear scan of the join table.  Given the typical amounts of data in a client application and the speed of Javascript engines, an O(n) operation is likely to have sufficiently fast performance for most use cases
```

Comment:
I'm not sure if it's a good idea to use big-O notation here, because I'm not sure how much of the audience would understand it without further explanation.  It's probably OK to leave it at  "linear scan" and people would understand what you mean.